### PR TITLE
add strategies to transform logs to subgraph

### LINKED ISSUE #215: add transformer strategy that parses from block-logs transformation to NFT ID (compatible with web3subgraph)
- see this snippet https://github.com/neume-network/strategies/blob/3b0480d9c5e0a39e3f87cea410fc19207f788794/src/strategies/call-block-logs/transformer.mjs#L77
- make it a separate strategy
- blocked by #177

### PR DESCRIPTION
**Pull request by il3ven**
This PR modifies `call-block-logs` and adds two new strategies `filter-music-nfts` and `logs-to-subgraph`. Closes #215.

After much thought I have settled for the following approach. It isn't perfect but it is the best. Let me explain the flow I am proposing.

**Flow:**
- Execute `call-block-logs` extractor to fetch all logs
- Execute `filter-music-nfts` extractor to extract all music NFT addresses. We can hand maintain addresses for zora, catalog, mint and zoid. For sound we need to extract them from the logs. This strategy does exactly that. It outputs a list of addresses that are music NFTs. Example:
  ```
  {
    "0xabefbc9fd2f806065b4f3c237d4b59d9a97bcac7": {
      name: "zora",
    },
    "0x0bc2a24ce568dad89691116d5b34deb6c203f342": {
      name: "catalog",
      version: "2.0.0",
    },
    "0xf5819e27b9bad9f97c177bf007c1f96f26d91ca6": {
      name: "noizd",
    },
    "0x2b5426a5b98a3e366230eba9f95a24f09ae4a584": {
      name: "mintsongs",
      version: "2.0.0",
    },
    "0xca13eaa6135d719e743ffebb5c26de4ce2f9600c": {
      "name": "sound"
    }
  }
  ```
- Execute `call-block-logs` transformer with a list of contracts and events to monitor. The above generated list is passed to `call-block-logs` as input.
- Execute `logs-to-subgraph` transformer to transform `call-block-logs` output to `web3subgraph` compatible output.

**Pros:**
- Keep `call-block-logs` generic as per #210.
- To monitor more platforms we have to only modify `filter-music-nfts`. It has neume specific logic.

**Cons:**
- Number of strategies has increased which can feel a little polluted
- We have to filter all logs twice. Once in `filter-music-nfts` and once in `call-block-logs` transformer.

**Done:**
- [x] Verified correctness by running it locally with logs from 47764 blocks
- [x] Add tests

**Can be done later:**
- [ ] https://github.com/neume-network/strategies/issues/219

<details>
<summary>Used the following crawl path</summary>

```js
import { resolve } from "path";
import { env } from "process";

// TODO: Remove this from here
const contracts = {
  "0xabefbc9fd2f806065b4f3c237d4b59d9a97bcac7": {
    name: "zora",
  },
  "0x0bc2a24ce568dad89691116d5b34deb6c203f342": {
    name: "catalog",
    version: "2.0.0",
  },
  "0xf5819e27b9bad9f97c177bf007c1f96f26d91ca6": {
    name: "noizd",
  },
  "0x2b5426a5b98a3e366230eba9f95a24f09ae4a584": {
    name: "mintsongs",
    version: "2.0.0",
  },
};

const events = ["Transfer(address,address,uint256)"];

// crawlPath[i] and crawlPath[i+1] are executed in sequence
// crawlPath[i][j] and crawlPath[i][j+1] are executed in parallel
export default [
  [
    {
      name: "call-block-logs",
      extractor: {
        args: [15232606,  15280370],
      },
    },
  ],
  [
    {
      name: "filter-music-nfts",
      extractor: {
        args: [resolve(env.DATA_DIR, "call-block-logs-extraction"), contracts],
      },
    },
  ],
  [
    {
      name: "call-block-logs",
      transformer: {
        args: [
          resolve(env.DATA_DIR, "call-block-logs-extraction"),
          resolve(env.DATA_DIR, "filter-music-nfts-extraction"),
          events
        ],
      },
    },
  ],
  [
    {
      name: "logs-to-subgraph",
      transformer: {
        args: [
          resolve(env.DATA_DIR, "call-block-logs-transformation"),
          resolve(env.DATA_DIR, "filter-music-nfts-extraction"),
        ],
      },
    },
  ],
  [
    {
      name: "soundxyz-call-tokenuri",
      extractor: {
        args: [resolve(env.DATA_DIR, "logs-to-subgraph-transformation")],
      },
      transformer: {},
    },
    {
      name: "zora-call-tokenuri",
      extractor: {
        args: [resolve(env.DATA_DIR, "logs-to-subgraph-transformation")],
      },
      transformer: {},
    },
    {
      name: "zora-call-tokenmetadatauri",
      extractor: {
        args: [resolve(env.DATA_DIR, "logs-to-subgraph-transformation")],
      },
      transformer: {},
    },
  ],
  [
    {
      name: "soundxyz-get-tokenuri",
      extractor: {
        args: [resolve(env.DATA_DIR, "soundxyz-call-tokenuri-transformation")],
      },
      transformer: {},
    },
    {
      name: "zora-get-tokenuri",
      extractor: {
        args: [
          resolve(env.DATA_DIR, "zora-call-tokenmetadatauri-transformation"),
        ],
      },
      transformer: {},
    },
  ],
  [
    {
      name: "music-os-accumulator",
      extractor: { args: [] }
    },
  ],
];
```

</details>

### CODE REVIEW

File: src/strategies/call-block-logs/extractor.mjs [hunk 1]
```diff
@@ -76,15 +76,8 @@ export function update(message) {
   if (message?.method == "eth_blockNumber") {
     let endBlockNumber = message.metadata.end;
     const latestBlockNumber = parseInt(message.results);
-    if (
```
Comment (il3ven) [inline src/strategies/call-block-logs/extractor.mjs:79]: Changed this because it seems like a bug.

File: src/strategies/call-block-logs/transformer.mjs [hunk 2]
```diff
@@ -23,84 +23,41 @@ export function onError(error) {
...
-  );
-  return result.artistAddress.toLowerCase();
-}
+let validatedContracts;
 
 const emptyB32 =
   "0x0000000000000000000000000000000000000000000000000000000000000000";
 
-export function onLine(line) {
+export function onLine(line, contractsPath, events) {
```
Comment (il3ven) [inline src/strategies/call-block-logs/transformer.mjs:38]: Ideally, I don't want to use `contractsPath` here but `filter-music-nfts` generates this and we can't simply import that file.

File: src/strategy-factories/call-tokenuri/extractor.mjs [hunk 3]
```diff
@@ -66,7 +70,7 @@ export const callTokenUriFactory = (props) => {
           to: address,
           data,
         },
-        toHex(BLOCK_NUMBER),
+        toHex(Math.max(parseInt(blockNumber), BLOCK_NUMBER)),
```
Comment (il3ven) [inline src/strategy-factories/call-tokenuri/extractor.mjs:73]: For NFTs minted after `BLOCK_NUMBER` we will have to use a higher block number. Perhaps the one coming from `createdAtBlockNumber`.

File: utils/snapshot_extractor.mjs [hunk 4]
```diff
@@ -54,13 +54,16 @@ export default async function snapshotExtractor(extractor, { inputs }) {
...
+  );
+
+  if (preparedMessages.length) {
+    preparedMessages.forEach((message) => {
       postMessage(message);
     });
-
-  await once(worker, "exit");
+    await once(worker, "exit");
+  }
```
Comment (il3ven) [inline utils/snapshot_extractor.mjs:67]: Fixed a bug that if `init` returns `messages: []` the `snapshotExtractor` doesn't close.

**Review by il3ven — COMMENTED** [review]
Added a few comments to explain the context behind the change.

File: src/strategies/call-block-logs/extractor.mjs (hunk 1 above)
Comment (TimDaub) [inline src/strategies/call-block-logs/extractor.mjs:79]: Please exclude this change from the PR. And my preference in general is always to neatly separate concerns. This makes the entire PR very hard to approve quickly.

File: src/strategies/filter-music-nfts/extractor.mjs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        indexed: true,
+      },
+    ],
+    log.data,
+    topics
+  );
+  return result.artistAddress.toLowerCase();
+}
+
+export async function init(logPath, contracts) {
```
Comment (TimDaub) [inline src/strategies/filter-music-nfts/extractor.mjs:43]: we must not abuse the extractor interface for filtering

File: src/strategy-factories/call-tokenuri/extractor.mjs (hunk 3 above)
Comment (TimDaub) [inline src/strategy-factories/call-tokenuri/extractor.mjs:73]: this change has nothing to do with block logs or filtering it must not be included in this pr. I'm stopping review here and close

**Comment by TimDaub** [comment]
please separate concerns accordingly. One problem, one issue, one commit one PR.

**Comment by TimDaub** [comment]
Assigning myself to separate concerns.

File: utils/snapshot_extractor.mjs (hunk 4 above)
Comment (TimDaub) [reply utils/snapshot_extractor.mjs:67]: should have been a separate pr and an issue

Comment (TimDaub) [reply utils/snapshot_extractor.mjs:67]: https://github.com/neume-network/strategies/commit/9ee5c6d2e4b9f8ac734c4650fc7226c1279c1a22

File: src/strategy-factories/call-tokenuri/extractor.mjs (hunk 3 above)
Comment (TimDaub) [reply src/strategy-factories/call-tokenuri/extractor.mjs:73]: https://github.com/neume-network/strategies/commit/edc666543c6160a334f702ec68a79b2c3f3dea67

File: src/strategies/logs-to-subgraph/transformer.mjs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+export function onError(error) {
+  log(error.toString());
+  throw error;
+}
+
+let validatedContracts;
+
+export function onLine(line, contractsPath) {
+  const contracts = JSON.parse(fs.readFileSync(contractsPath));
```
Comment (TimDaub) [inline src/strategies/logs-to-subgraph/transformer.mjs:27]: considering that this function is called very often, doing a readFileSync is every time is inefficient

File: test/strategies/logs-to-subgraph/transformer_test.mjs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+test("soundxyz-call-tokenuri transformer", (t) => {
```
Comment (TimDaub) [inline test/strategies/logs-to-subgraph/transformer_test.mjs:15]: this test should not be named "soundxyz-call-tokenuri transformed"

**Comment by il3ven** [comment]
> Assigning myself to separate concerns.

I'll try to be better next time. 😓

File: src/strategies/logs-to-subgraph/transformer.mjs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  } catch (err) {
+    return {
+      write: null,
+      messages: [],
+    };
+  }
+
+  logs = logs.map((log) => ({
+    address: log.address,
+    tokenId: `${BigInt(log.topics[3]).toString(10)}`,
```
[the remark this answers is not in the document]
Comment (TimDaub) [reply src/strategies/logs-to-subgraph/transformer.mjs:56]: my bad, this leads to a bug when log.topics[3] is bigger than Number.MAX_INT

[the remark this answers is not in the document]
Comment (il3ven) [reply src/strategies/logs-to-subgraph/transformer.mjs:56]: I should have explained this. A few tokenIds are too big to fit in Number. Check this https://etherscan.io/tx/0xd3338f56c566a9278e5218f3797a8fa84f113654565be2c25cf97e3599959ad3#eventlog.

`parseInt('0x0000000000000000000000000000000200000000000000000000000000000007', 16)` gives us `6.80564733841877e+38`.

`BigInt('0x0000000000000000000000000000000200000000000000000000000000000007').toString(10)` gives us `680564733841876926926749214863536422919`.

File: src/strategies/logs-to-subgraph/transformer.mjs (hunk 6 above)
Comment (il3ven) [reply src/strategies/logs-to-subgraph/transformer.mjs:27]: I agree. It is not efficient but since `onLine` isn't async I couldn't use `fs.readFile`.

File: src/strategies/logs-to-subgraph/transformer.mjs (hunk 8 above)
[the remark this answers is not in the document]
Comment (TimDaub) [reply src/strategies/logs-to-subgraph/transformer.mjs:56]: this is a great catch, thank you and sorry for being so harsh on that one.

File: src/strategies/filter-music-nfts/extractor.mjs (hunk 5 above)
Comment (il3ven) [reply src/strategies/filter-music-nfts/extractor.mjs:43]: Should we use transformer interface for filtering? My thought was that it is extracting a subset of data from the larger dataset.

Comment (TimDaub) [reply src/strategies/filter-music-nfts/extractor.mjs:43]: @il3ven I'd like to understand: What stopped you from using the transformer interface to handle this type of task? If we need a "filter" strategy interface (that is different from transform and extraction, we should consider adding it.

Comment (TimDaub) [reply src/strategies/filter-music-nfts/extractor.mjs:43]: > Should we use transformer interface for filtering? My thought was that it is extracting a subset of data from the larger dataset.

mhh language-wise that makes sense to me but I I think the extractor is supposed to be used to download data from the internet.

**Comment by TimDaub** [comment]
I have separated this PR into three commits:

- edc666543c6160a334f702ec68a79b2c3f3dea67 introduces a minimal block number for call-tokenuri 
- 9ee5c6d2e4b9f8ac734c4650fc7226c1279c1a22 fixes a bug in the utils/snapshotextractor
- 271e27cf379591728d14934794428c595f4f065d: parses event logs into web3subgraph NFT IDs

There is one leftover question in 271e27cf379591728d14934794428c595f4f065d, which is how we should match "platform" in cases like sound's custom contracts.

File: src/strategies/filter-music-nfts/extractor.mjs (hunk 5 above)
Comment (il3ven) [reply src/strategies/filter-music-nfts/extractor.mjs:43]: > What stopped you from using the transformer interface to handle this type of task?

Nothing. language-wise it seemed natural to use extractor. I am fine with transformer too.

**Comment by il3ven** [comment]
> There is one leftover question in https://github.com/neume-network/strategies/commit/271e27cf379591728d14934794428c595f4f065d, which is how we should match "platform" in cases like sound's custom contracts.

Yes, this is the problem that lead to the not so ideal approach of filter-music-nfts.

**Comment by TimDaub** [comment]
> Yes, this is the problem that lead to the not so ideal approach of filter-music-nfts.

I see and actually it is somewhat consistent to our approach with e.g. what we do at music-os-accumulator. I've submitted a failing test as an PR that exemplifies the issue: https://github.com/neume-network/strategies/pull/235
